File: selenium/webdriver-framework-core/src/main/java/com/browserstack/webdriver/core/LocalFactory.java
package com.browserstack.webdriver.core;

import com.browserstack.local.Local;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

public class LocalFactory {

    private static final Logger LOGGER = LoggerFactory.getLogger(LocalFactory.class);

    private static volatile LocalFactory instance;

    private final Local local = new Local();
    private final String localIdentifier;

    private LocalFactory(Map<String, String> localOptions) {
        try {
            localIdentifier = localOptions.containsKey("localIdentifier")?localOptions.get("localIdentifier"):RandomStringUtils.randomAlphabetic(8);
            localOptions.put("localIdentifier", localIdentifier);
            if (System.getProperty("https.proxyHost") != null) {
                setProxy("https", localOptions);
            } else {
                if (System.getProperty("http.proxyHost") != null) {
                    setProxy("http", localOptions);
                }
            }
            local.start(localOptions);
            LOGGER.debug("Started BrowserStack Local with identifier {}.", localIdentifier);
        } catch (Exception e) {
            LOGGER.error("Initialization of BrowserStack Local failed.");
            throw new RuntimeException("Initialization of BrowserStack Local failed.", e);
        }
    }

    private void setProxy(String protocol, Map<String, String> localOptions) {
        String proxyHost = System.getProperty(String.format("%s.proxyHost", protocol));
        String proxyPort = System.getProperty(String.format("%s.proxyPort", protocol));
        String proxyUser = System.getProperty(String.format("%s.proxyUser", protocol));
        String proxyPassword = System.getProperty(String.format("%s.proxyPassword", protocol));
        String proxyHostPrefix = "proxy";
        String proxyUserPrefix = "proxy";
        if (proxyHost.startsWith("local") || proxyHost.startsWith("127.")) {
            proxyHostPrefix = "localProxy";
            proxyUserPrefix = "-localProxy";
        }
        localOptions.put(String.format("%sHost", proxyHostPrefix), proxyHost);
        localOptions.put(String.format("%sPort", proxyHostPrefix), proxyPort);
        if (StringUtils.isNotEmpty(proxyUser)) {
            localOptions.put(String.format("%sUser", proxyUserPrefix), proxyUser);
            localOptions.put(String.format("%sPass", proxyUserPrefix), proxyPassword);
        }
        LOGGER.debug("Proxy configurations detected Host : {}, Port : {}, User: {}.", proxyHost, proxyPort, proxyUser);
    }

    public static void createInstance(Map<String, String> args) {
        if (instance == null) {
            synchronized (LocalFactory.class) {
                if (instance == null) {
                    instance = new LocalFactory(args);
                    Runtime.getRuntime().addShutdownHook(new Closer(instance.local));
                }
            }
        }
    }

    public static LocalFactory getInstance() {
        return instance;
    }

    public String getLocalIdentifier() {
        return instance.localIdentifier;
    }

    private static class Closer extends Thread {
        private final Local local;

        public Closer(Local local) {
            this.local = local;
        }

        @Override
        public void run() {
            try {
                if (local.isRunning()) {
                    local.stop();
                }
            } catch (Exception e) {
               LOGGER.error("Error encountered while stopping BrowserStack Local { }",e);
            }
        }
    }
}

File: webdriver-framework-core-python/src/main/webdriver/core/LocalFactory.py
from browserstack.local import Local
import os
import atexit
import threading
import string
import random

class LocalFactoryClass:
    instance = None 

    local = Local()
    localIdentifier : str = ""

    LocalFactoryClassLock = threading.Lock()


    def __init__(self, localOptions) :
        try :
            localIdentifier = ''.join(random.choices(string.ascii_uppercase + string.digits, k = 7))
            if "localIdentifier" in localOptions :
                localIdentifier = localOptions["localIdentifier"]

            localOptions["localIdentifier"] =  localIdentifier

            if (os.getenv("https.proxyHost") != None) :
                self.setProxy("https", localOptions)
            else:
                if (os.getenv("http.proxyHost") != None) :
                    self.setProxy("http", localOptions)

            self.local.start(**localOptions)
            LocalFactoryClass.localIdentifier = localIdentifier

            print(f"Started BrowserStack Local with identifier {localIdentifier}.")
        except Exception as e :
            print("Initialization of BrowserStack Local failed.")
            print(e)
            # throw new RuntimeException("Initialization of BrowserStack Local failed.", e);
    
    def setProxy(protocol:str, localOptions: dict) :
        proxyHost = os.getenv(protocol+".proxyHost")
        proxyPort = os.getenv(protocol+".proxyPort")
        proxyUser = os.getenv(protocol+".proxyUser")
        proxyPassword = os.getenv(protocol+".proxyPassword")
        proxyHostPrefix = "proxy"
        proxyUserPrefix = "proxy"
        if (proxyHost.startsWith("local") or proxyHost.startsWith("127.")):
            proxyHostPrefix = "localProxy"
            proxyUserPrefix = "-localProxy"
        
        localOptions[proxyHostPrefix+"Host"] = proxyHost
        localOptions[proxyHostPrefix+"Port"] =  proxyPort
        if (proxyUser != None or proxyUser != "") :
            localOptions[proxyUserPrefix+ "User"] = proxyUser
            localOptions[proxyUserPrefix+"Pass"] = proxyPassword
    
        print(f"Proxy configurations detected Host : {proxyHost}, Port : {proxyPort}, User: {proxyUser}." )

    @staticmethod
    def createInstance(args:dict) :
        #Make thread safe
        if (LocalFactoryClass.instance == None) :
            with LocalFactoryClass.LocalFactoryClassLock:
                if (LocalFactoryClass.instance == None) :
                    LocalFactoryClass.instance = LocalFactoryClass(args)
                    print("Starting Local Instance...")
                    atexit.register(LocalFactoryClass.stopLocal)
            
    @staticmethod
    def getInstance() :
        return LocalFactoryClass.instance

    @staticmethod
    def getLocalIdentifier():
        return LocalFactoryClass.instance.localIdentifier

    @staticmethod
    def stopLocal():
        if (LocalFactoryClass.instance.local.isRunning()):
            print("Stopping Local Instance")
            LocalFactoryClass.instance.local.stop()
        else:
            print("Local is not running [IGNORE]")
        